Library/Homebrew/cask/lib/hbc/cask.rb

```diff
@@ -10,6 +10,13 @@ class Cask
     def initialize(token, sourcefile_path: nil, &block)
       @token = token
       @sourcefile_path = sourcefile_path
+
+      @tap = if @sourcefile_path == CaskLoader.default_path(@token)
+        Hbc.default_tap
+      elsif @sourcefile_path.to_s =~ HOMEBREW_TAP_PATH_REGEX
+        Tap.fetch(Regexp.last_match(1), Regexp.last_match(2))
+      end
```

Comment:
I found this to be required and correct for third-party taps. For example, the `sourcefile_path` for caskroom/homebrew-versions/microsoft-remote-desktop-beta is "/usr/local/Homebrew/Library/Taps/caskroom/homebrew-versions/Casks/microsoft-remote-desktop-beta.rb" in my environment. `Regexp.last_match(1)` is "caskroom" and `Regexp.last_match(2)` is "homebrew-versions".